### PR TITLE
fix: TimeoutTicker returns wrong value/timeout pair when timeouts are scheduled at ~approximately the same time

### PR DESCRIPTION
**Pull request by ValarDragon**
#3091 

The problem is we have an edge case where we should drain the timer channel, but we "let it slide" in certain race conditions when two timeouts are scheduled near each other. This means we can have unsafe timeout behavior as demonstrated in the github issue, and likely more spots in consensus.

Notice that aside from NewTimer and OnStop, all timer accesses are from the same thread. In NewTimer we can block until the timer is drained (very quickly up to goroutine scheduling). In OnStop we don't need to guarantee draining before the method ends, we can just launch something into the channel that will kill it.

In the main timer goroutine, we can safely maintain this "timerActive" variable, and force drain when its active. This removes the edge case.

The test I created does fail on main.

---

#### PR checklist

- [X] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [X] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: internal/consensus/ticker.go [hunk 1]
```diff
@@ -29,17 +29,23 @@ type TimeoutTicker interface {
...
-		timer:    time.NewTimer(0),
-		tickChan: make(chan timeoutInfo, tickTockBufferSize),
-		tockChan: make(chan timeoutInfo, tickTockBufferSize),
+		timer: time.NewTimer(0),
+		// An indicator variable to check if the timer is active or not.
+		// Concurrency safe because the timer is only accessed by a single goroutine.
+		timerActive: true,
+		tickChan:    make(chan timeoutInfo, tickTockBufferSize),
+		tockChan:    make(chan timeoutInfo, tickTockBufferSize),
+		stopChan:    make(chan struct{}),
```
Comment (melekes) [inline internal/consensus/ticker.go:48]: don't think we need it. after `OnStop` is called, `t.Quit()` is closed.

Comment (ValarDragon) [reply internal/consensus/ticker.go:48]: Thank you! Wasn't familiar with thats how stop works, I'll remove that

**Review by melekes — APPROVED** [review]
Thanks @ValarDragon ❤️

File: internal/consensus/ticker.go [hunk 2]
```diff
@@ -29,17 +29,21 @@ type TimeoutTicker interface {
...
-		tockChan: make(chan timeoutInfo, tickTockBufferSize),
+		timer: time.NewTimer(0),
+		// An indicator variable to check if the timer is active or not.
+		// Concurrency safe because the timer is only accessed by a single goroutine.
+		timerActive: true,
+		tickChan:    make(chan timeoutInfo, tickTockBufferSize),
+		tockChan:    make(chan timeoutInfo, tickTockBufferSize),
 	}
 	tt.BaseService = *service.NewBaseService(nil, "TimeoutTicker", tt)
 	tt.stopTimer() // don't want to fire until the first scheduled timeout
```
Comment (hvanz) [inline internal/consensus/ticker.go:49]: Since timerActive is true, stopTimer will return immediately, so this line could be removed?

**Review by hvanz — APPROVED** [review]
LGTM

File: internal/consensus/ticker_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+		// Wait for the timeout to be received
+		to := <-c
+		endTime := time.Now()
+		elapsedTime := endTime.Sub(startTime)
+		if timeout == to {
+			require.True(t, elapsedTime >= timeout.Duration, "We got the 5ms timeout. However the timeout happened too quickly. Should be >= 5ms. Got %dms (start time %d end time %d)", elapsedTime.Milliseconds(), startTime.UnixMilli(), endTime.UnixMilli())
+		} else {
+			require.True(t, elapsedTime <= timeout.Duration, "We got the -1ms timeout. However the timeout happened too slowly. Should be ~near instant. Got %dms (start time %d end time %d)", elapsedTime.Milliseconds(), startTime.UnixMilli(), endTime.UnixMilli())
+		}
```
Comment (sergio-mena) [inline internal/consensus/ticker_test.go:40]: Aren't the conditions reversed? E.g. `elapsedTime >= timeout.Duration` means "more time has elapsed than expected" so it happened too slowly. Same for the _else_.

File: internal/consensus/ticker_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+		ticker.ScheduleTimeout(timeout)
+
+		// Wait for the timeout to be received
+		to := <-c
+		endTime := time.Now()
+		elapsedTime := endTime.Sub(startTime)
+		if timeout == to {
+			require.True(t, elapsedTime >= timeout.Duration, "We got the 5ms timeout. However the timeout happened too quickly. Should be >= 5ms. Got %dms (start time %d end time %d)", elapsedTime.Milliseconds(), startTime.UnixMilli(), endTime.UnixMilli())
+		} else {
+			require.True(t, elapsedTime <= timeout.Duration, "We got the -1ms timeout. However the timeout happened too slowly. Should be ~near instant. Got %dms (start time %d end time %d)", elapsedTime.Milliseconds(), startTime.UnixMilli(), endTime.UnixMilli())
```
Comment (sergio-mena) [inline internal/consensus/ticker_test.go:39]:
```suggestion
			require.True(t, elapsedTime <= to.Duration, "We got the -1ms timeout. However the timeout happened too slowly. Should be ~near instant. Got %dms (start time %d end time %d)", elapsedTime.Milliseconds(), startTime.UnixMilli(), endTime.UnixMilli())
```

**Review by sergio-mena — APPROVED** [review]
Excellent! Thanks @ValarDragon 🙏

**Comment by ValarDragon** [comment]
@sergio-mena re your comment yes, but the timeout here used is 0s, which is supposed to execute ~near instantly (whereas its budgeted 5ms). I believe this has different scheduler logic than things budgeted further into the future. 

Will remove though, since as you note its covered by the then branch regardless!

**Comment by cason** [comment]
Hey, just to mention that timeout scheduling/events are totally ordered. That is why when we receive a command to schedule timeout for a height (H), round (S), and step (RS), we compare it to the currently scheduled. If it is "larger", we replace the old timeout by the new one. This ad-hoc solution prevents a lot of timeouts to expire when they are not needed anymore.

In your test case you use two timeouts that are equal. This implementation does not support this, because of the total order assumption, that is why the test fails.
